File: references/tools/serve.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""松鼠大战怀旧复刻版 —— 纯 Python 本地服务器（不需要 Node.js）

用法:
    python3 references/tools/serve.py [端口]        # 默认 8080
    python3 references/tools/serve.py 8080 --no-save

和 references/tools/serve.js 行为一致：静态文件 + `/__save` 存档接口
（存档写在 <游戏目录>/save/progress.json）。macOS 自带 python3，所以在
没有 Node.js 的机器上也能把进度存进文件，而不是退回浏览器 localStorage。
"""
import json
import os
import sys
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import urlparse, parse_qs

ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
SAVE_DIR = os.path.join(ROOT, "save")
SAVE_FILE = os.path.join(SAVE_DIR, "progress.json")
SAVE_MAX = 4 * 1024 * 1024
ARGS = sys.argv[1:]
NO_SAVE = "--no-save" in ARGS
PORTS = [int(a) for a in ARGS if a.isdigit()]
PORT = PORTS[0] if PORTS else 8080


def save_meta():
    if not os.path.exists(SAVE_FILE):
        return {"exists": False, "savedAt": 0, "size": 0}
    st = os.stat(SAVE_FILE)
    return {"exists": True, "savedAt": int(st.st_mtime * 1000), "size": st.st_size}


class Handler(SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=ROOT, **kwargs)

    def log_message(self, fmt, *args):        # 静音：只在出错时自己打印
        pass

    def _json(self, code, body):
        data = json.dumps(body, ensure_ascii=False).encode("utf-8")
        self.send_response(code)
        self.send_header("content-type", "application/json; charset=utf-8")
        self.send_header("cache-control", "no-store")
        self.send_header("content-length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def _save_api(self, query):
        if NO_SAVE:
            return self._json(503, {"ok": False, "msg": "服务器以 --no-save 启动，存档文件同步已关闭"})
        meta = save_meta()
        if self.command == "GET":
            if "meta" in query:
                return self._json(200, {"ok": True, "path": SAVE_FILE, **meta})
            if not meta["exists"]:
                return self._json(200, {"ok": True, "exists": False, "savedAt": 0, "data": None, "path": SAVE_FILE})
            with open(SAVE_FILE, "r", encoding="utf-8") as fh:
                text = fh.read()
            return self._json(200, {"ok": True, "exists": True, "savedAt": meta["savedAt"], "data": text, "path": SAVE_FILE})
        if self.command == "POST":
            length = int(self.headers.get("content-length") or 0)
            if length <= 0 or length > SAVE_MAX:
                return self._json(413, {"ok": False, "msg": "存档太大或读取中断"})
            raw = self.rfile.read(length).decode("utf-8")
            try:
                parsed = json.loads(raw)
            except ValueError:
                return self._json(400, {"ok": False, "msg": "不是合法 JSON"})
            if not isinstance(parsed, dict):
                return self._json(400, {"ok": False, "msg": "存档必须是对象"})
            os.makedirs(SAVE_DIR, exist_ok=True)
            with open(SAVE_FILE, "w", encoding="utf-8") as fh:
                json.dump(parsed, fh, ensure_ascii=False)
            return self._json(200, {"ok": True, "savedAt": save_meta()["savedAt"], "path": SAVE_FILE})
        return self._json(405, {"ok": False, "msg": "只支持 GET / POST"})

    def do_GET(self):
        parsed = urlparse(self.path)
        if parsed.path == "/__save":
            return self._save_api(parse_qs(parsed.query))
        return super().do_GET()

    def do_POST(self):
        parsed = urlparse(self.path)
        if parsed.path == "/__save":
            return self._save_api(parse_qs(parsed.query))
        self.send_error(404, "not found")


if __name__ == "__main__":
    print("松鼠大战 · 怀旧复刻版（Python 服务器）")
    print("目录: " + ROOT)
    print("地址: http://127.0.0.1:%d/" % PORT)
    print("自检: http://127.0.0.1:%d/index.html?test=1" % PORT)
    print("存档文件: " + ("已关闭（--no-save）" if NO_SAVE else SAVE_FILE))
    ThreadingHTTPServer(("127.0.0.1", PORT), Handler).serve_forever()

File: scripts/serve.js
#!/usr/bin/env node
/* ============================================================
 * tools/serve.js — 本地静态服务器（零依赖）
 *
 * 用法: node tools/serve.js [端口] [--no-save]
 * 然后浏览器打开 http://127.0.0.1:8080
 *
 * 说明：直接双击 index.html（file://）也能玩，但部分浏览器会
 * 拦截本地文件读取，用这个服务器最稳。
 *
 * 存档文件（可选）：游戏会把存档同时写一份到 <游戏目录>/save/progress.json，
 * 既可当本地备份，也是接「自建同步服务」的接口层。
 * 接口只监听 127.0.0.1，加 --no-save 可整体关闭。
 * ============================================================ */
const http = require('http');
const fs = require('fs');
const path = require('path');

const ROOT = path.resolve(__dirname, '..');   // serve.js 在 scripts/ 里，游戏根是上一层
const ARGS = process.argv.slice(2);
const PORT = Number(ARGS.find((a) => /^[0-9]+$/.test(a)) || 8080);   // 端口可以写在任意位置：node tools/serve.js --no-save 8080

/* ---------- /__save：存档文件读写（本地备份 / 自建同步服务的接口层） ---------- */
const NO_SAVE = process.argv.includes('--no-save');
const SAVE_DIR = path.join(ROOT, 'save');
/* 首页 index.html 住在 scripts/ 里。浏览器地址仍然是 http://host:port/ （或 /index.html），
 * 页面里的 css/js/images/audio 都是相对 URL 根解析的，所以 html 里一个引用都不用改。 */
const ENTRY_REL = ['scripts/index.html', 'index.html'].find((f) => fs.existsSync(path.join(ROOT, f))) || 'index.html';
const SAVE_FILE = path.join(SAVE_DIR, 'progress.json');
const SAVE_MAX = 4 * 1024 * 1024;      // 存档上限 4 MB

function saveStat() {
  try { const st = fs.statSync(SAVE_FILE); return { exists: true, savedAt: Math.round(st.mtimeMs), size: st.size }; }
  catch (e) { return { exists: false, savedAt: 0, size: 0 }; }
}
function readBody(req, limit) {
  return new Promise((resolve, reject) => {
    let size = 0; const chunks = [];
    req.on('data', (c) => {
      size += c.length;
      if (size > limit) { reject(new Error('too large')); req.destroy(); return; }
      chunks.push(c);
    });
    req.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
    req.on('error', reject);
  });
}
function json(res, code, body) {
  res.writeHead(code, { 'content-type': 'application/json; charset=utf-8', 'cache-control': 'no-store' });
  res.end(JSON.stringify(body));
}
async function handleSaveApi(req, res, url) {
  if (NO_SAVE) return json(res, 503, { ok: false, msg: '服务器以 --no-save 启动，存档文件同步已关闭' });
  if (req.method === 'GET') {
    const st = saveStat();
    if (url.searchParams.get('meta')) return json(res, 200, { ok: true, exists: st.exists, savedAt: st.savedAt, size: st.size });
    if (!st.exists) return json(res, 200, { ok: true, exists: false, savedAt: 0, data: null });
    try { return json(res, 200, { ok: true, exists: true, savedAt: st.savedAt, data: fs.readFileSync(SAVE_FILE, 'utf8') }); }
    catch (e) { return json(res, 500, { ok: false, msg: '读取存档失败：' + e.message }); }
  }
  if (req.method === 'POST') {
    let body;
    try { body = await readBody(req, SAVE_MAX); }
    catch (e) { return json(res, 413, { ok: false, msg: '存档太大或读取中断' }); }
    let parsed;
    try { parsed = JSON.parse(body); } catch (e) { return json(res, 400, { ok: false, msg: '不是合法 JSON' }); }
    if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) return json(res, 400, { ok: false, msg: '存档必须是对象' });
    try {
      fs.mkdirSync(SAVE_DIR, { recursive: true });
      fs.writeFileSync(SAVE_FILE, JSON.stringify(parsed), 'utf8');
      return json(res, 200, { ok: true, savedAt: saveStat().savedAt });
    } catch (e) { return json(res, 500, { ok: false, msg: '写入存档失败：' + e.message }); }
  }
  return json(res, 405, { ok: false, msg: '只支持 GET / POST' });
}

const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.webp': 'image/webp',
  '.gif': 'image/gif',
  '.svg': 'image/svg+xml',
  '.mp3': 'audio/mpeg',
  '.mp4': 'video/mp4',
  '.json': 'application/json; charset=utf-8',
  '.ico': 'image/x-icon',
};

http.createServer((req, res) => {
  let rel, url;
  try {
    url = new URL(req.url, 'http://127.0.0.1');
    rel = decodeURIComponent(url.pathname);
  } catch (_) { res.writeHead(400); res.end('400'); return; }
  if (rel === '/__save') { handleSaveApi(req, res, url).catch(() => json(res, 500, { ok: false, msg: '服务器内部错误' })); return; }
  if (rel === '/' || rel === '' || rel === '/index.html') rel = '/' + ENTRY_REL;
  const file = path.join(ROOT, rel);
  if (file !== ROOT && !file.startsWith(ROOT + path.sep)) { res.writeHead(403); res.end('403'); return; }
  fs.readFile(file, (err, data) => {
    if (err) { res.writeHead(404, { 'content-type': 'text/plain; charset=utf-8' }); res.end('404 ' + rel); return; }
    res.writeHead(200, {
      'content-type': MIME[path.extname(file).toLowerCase()] || 'application/octet-stream',
      'cache-control': 'no-cache',
    });
    res.end(data);
  });
}).listen(PORT, '127.0.0.1', () => {
  const st = saveStat();
  console.log('松鼠大战 · 怀旧复刻版');
  console.log('目录: ' + ROOT);
  console.log('地址: http://127.0.0.1:' + PORT + '/');
  console.log('自检: http://127.0.0.1:' + PORT + '/index.html?test=1');
  console.log('战斗观察: http://127.0.0.1:' + PORT + '/index.html?test=2');
  console.log(NO_SAVE
    ? '存档文件: 已关闭（--no-save）'
    : '存档文件: save/progress.json' + (st.exists ? '（已有存档 ' + new Date(st.savedAt).toLocaleString() + '）' : '（还没有，游戏保存后自动出现）'));
});
